proxy/http2/Http2Stream.cc

```diff
@@ -190,22 +191,34 @@ Http2Stream::main_event_handler(int event, void *edata)
   case VC_EVENT_READ_COMPLETE:
   case VC_EVENT_READ_READY:
     _timeout.update_inactivity();
-    if (e->cookie == &read_vio) {
+    if (e && e->cookie == &read_vio) {
       if (read_vio.mutex && read_vio.cont && this->_sm) {
         signal_read_event(event);
       }
     } else {
       this->update_read_request(true);
     }
     break;
+  case VC_EVENT_ERROR:
   case VC_EVENT_EOS:
-    if (e->cookie == &read_vio) {
+    if (e && e->cookie == &read_vio) {
       SCOPED_MUTEX_LOCK(lock, read_vio.mutex, this_ethread());
-      read_vio.cont->handleEvent(VC_EVENT_EOS, &read_vio);
-    } else if (e->cookie == &write_vio) {
+      read_vio.cont->handleEvent(event, &read_vio);
+    } else if (e && e->cookie == &write_vio) {
       SCOPED_MUTEX_LOCK(lock, write_vio.mutex, this_ethread());
-      write_vio.cont->handleEvent(VC_EVENT_EOS, &write_vio);
+      write_vio.cont->handleEvent(event, &write_vio);
+    }
+
+    // This assuming stream on inbound side
+    // TODO: make appropriate state
+    if (recv_end_stream && write_vio.cont && write_vio.op == VIO::WRITE && write_vio.nbytes != 0) {
+      SCOPED_MUTEX_LOCK(lock, write_vio.cont->mutex, this_ethread());
+      write_vio.cont->handleEvent(event, &write_vio);
+    } else if (read_vio.cont && read_vio.op == VIO::READ) {
+      SCOPED_MUTEX_LOCK(lock, read_vio.cont->mutex, this_ethread());
+      read_vio.cont->handleEvent(event, &read_vio);
     }
```

Comment:
This second clause is causing our use after free.  Like 209 sends the EOS to our HttpTunnel which then causes everything to cleanup.  Then in line 215 we try to access the HttpSM via the write_vio.  Don't unnderstand why we need to do this two times.

```
  case VC_EVENT_ERROR:
  case VC_EVENT_EOS:
    if (e && e->cookie == &read_vio) {
      SCOPED_MUTEX_LOCK(lock, read_vio.mutex, this_ethread());
      read_vio.cont->handleEvent(event, &read_vio);
    } else if (e && e->cookie == &write_vio) {
      SCOPED_MUTEX_LOCK(lock, write_vio.mutex, this_ethread());
      write_vio.cont->handleEvent(event, &write_vio);
    }

    // This assuming stream on inbound side
    // TODO: make appropriate state
    if (recv_end_stream && write_vio.cont && write_vio.op == VIO::WRITE && write_vio.nbytes != 0) {
      SCOPED_MUTEX_LOCK(lock, write_vio.cont->mutex, this_ethread());
      write_vio.cont->handleEvent(event, &write_vio);
    } else if (read_vio.cont && read_vio.op == VIO::READ) {
      SCOPED_MUTEX_LOCK(lock, read_vio.cont->mutex, this_ethread());
      read_vio.cont->handleEvent(event, &read_vio);
    }
```

Reply:
Removing the second clause got rid of the quick use after free, but eventually a new core occurred.

Reply:
Thanks for testing! It looks like I haven't tested cases in which Http2Stream schedules events to itself.

> Don't unnderstand why we need to do this two times.

 Right, we need only one time. We need a guard of null check of `e` for the second block.

Reply:
f57bde4429f65dec7c0c3d2b03a6a190032b22e7 covers this (and probably the last crash on `HttpSM::do_http_server_open`).